docs/design.md

```diff
@@ -133,21 +198,65 @@ Checks if the required volume exists and is &quot;available&quot; (not attached
 - Checks that given volume is available (i.e. not attached to any other node) and returns error if it is attached.
 - Chooses the right device name for the volume on the node (more on that below) and issues AttachVolume. TODO: this has complicated idempotency expectations. It cancels previously called ControllerUnpublishVolume that may be still in progress (i.e. AWS is still detaching the volume and Kubernetes now wants the volume to be attached back).
 
+```mermaid
+sequenceDiagram
+    participant s as CSI Attacher
+    participant d as EBS Plugin
+    participant e as AWS API
+
+    s->>d: ControllerPublishVolume RPC
+    activate d
+    
+    d->>d: Parse Request + Ensure Idempotency
+
+    d->>e: EC2 DescribeInstances
+    
+    e-->>d: Get instanceID + all device names
+
+    d->>d: Assign likely unused device name
+    
+    d->>e: EC2 AttachVolume
+    activate e
+    e-->>d: Volume state == 'attaching'
+
+    d->>e: Poll EC2 DescribeVolumes
+    note over e: 2+ seconds
+    e-->>d: Volume state == 'in-use'
+    deactivate e
+    
+    d-->>s: Return device path 
+    deactivate d
+```
+
 #### ControllerUnpublishVolume
 
 Checks that given volume is not attached to given node. Returns success if so. Issues AWS.DetachVolume and marks the detached device name as free (more on that below). TODO: this has complicated idempotency expectations. It cancels previously called ControllerPublishVolume (i.e.AWS is still attaching the volume and Kubernetes now wants the volume to be detached).
 
+#### ControllerExpandVolume
+
+Checks that given volume is not expanded yet, calls EC2 ModifyVolume and ensures the modification enters the 'optimizing' state. 
+
+Note: If ControllerModifyVolume is triggered within 2 seconds of ControllerExpandVolume, they will share an EC2 ModifyVolume call. 
```

Comment:
Suggest saying `approximately 2 seconds` here as the way we do this isn't exact.